File: terminals/terminal.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <cstdio>
#include <memory>
#include <stdexcept>
#include <array>

std::string exec(const char* cmd) {
    std::array<char, 128> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
    if (!pipe) throw std::runtime_error("popen() failed!");
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        result += buffer.data();
    }
    return result;
}

void terminal_emulator() {
    // std::cout << "Python interpreter path: " << "N/A" << std::endl; // Путь к интерпретатору Python не доступен
    // std::cout << "Script path: " << "N/A" << std::endl; // Путь к скрипту не доступен

    std::vector<std::string> history;
    std::string command;

    while (true) {
        std::cout << ">> ";
        std::getline(std::cin, command);
        if (command == "exit" || command == "quit") {
            std::cout << "exiting..." << std::endl;
            break;
        }

        if (!command.empty()) {
            history.push_back(command);
        }

        try {
            std::string output = exec(command.c_str());
            std::cout << output;
        } catch (const std::exception& e) {
            std::cout << "Error: " << e.what() << std::endl;
        }
    }
}

void terminal_emulator_batch(const std::vector<std::string>& commands) {
    for (const auto& command : commands) {
        try {
            std::string output = exec(command.c_str());
            std::cout << output;
        } catch (const std::exception& e) {
            std::cout << "Error: " << e.what() << std::endl;
        }
    }
}

int main() {
    // terminal_emulator();
    terminal_emulator_batch({"ls -l"});
    return 0;
}

File: onelined-terminals/terminal.py
exec("""import subprocess
import sys
import os
import readline  # Импортируем библиотеку readline

def terminal_emulator():
    # Вывод путей к интерпретатору Python и к файлу программы
    print(f"Python interpreter path: {sys.executable}")
    print(f"Script path: {os.path.abspath(__file__)}")

    # Список для хранения истории команд
    history = []

    while True:
        command = input(">> ")
        if command.lower() in ['exit', 'quit']:
            print("exiting...")
            break
        
        # Добавляем команду в историю
        if command:
            history.append(command)
            readline.add_history(command)

        try:
            # Изменение на использование Popen для обработки ввода
            process = subprocess.Popen(command, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            
            # Чтение вывода в реальном времени
            while True:
                output = process.stdout.readline()
                if output == '' and process.poll() is not None:
                    break
                if output:
                    print(output.strip())
            
            # Обработка ошибок
            stderr = process.stderr.read()
            if stderr:
                print(f"Err: {stderr.strip()}")
                
        except Exception as e:
            print(f"Error: {e}")

def terminal_emulator_batch(commands):
    for command in commands:
        try:
            # Изменение на использование Popen для обработки ввода
            process = subprocess.Popen(command, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
            
            # Чтение вывода в реальном времени
            while True:
                output = process.stdout.readline()
                if output == '' and process.poll() is not None:
                    break
                if output:
                    print(output.strip())
            
            # Обработка ошибок
            stderr = process.stderr.read()
            if stderr:
                print(f"Err: {stderr.strip()}")
                
        except Exception as e:
            print(f"Error: {e}")

if __name__ == "__main__":
    terminal_emulator()
    # terminal_emulator_batch([
    #     "ifconfig",
    #     "ip addr",
    #     "curl -s \"myexternalip.com/raw\""
    # ])""")